File: src/main/java/com/malcolmcrum/spacetrader/GameManager.java
package com.malcolmcrum.spacetrader;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.malcolmcrum.spacetrader.GameStates.GameState;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Created by Malcolm on 9/10/2015.
 */
public class GameManager {
    private static final Logger logger = LoggerFactory.getLogger(GameManager.class);

    private Map<Integer, Game> games = new HashMap<>();
    private Map<Integer, GameState> states = new HashMap<>();

    public int newGame(String name, int fighter, int pilot, int trader, int engineer, int difficultyIndex) {
        Game game = new Game();
        games.put(game.id, game);
        Difficulty d = Difficulty.values()[difficultyIndex];
        states.put(game.id, game.startNewGame(name, fighter, pilot, trader, engineer, d));
        logger.info("New game created for Captain " + name + ", id: " + game.id);
        return game.id;
    }

    public Map<Integer, Game> getGames() {
        return games;
    }

    public GameState getState(int id) {
        if (states.containsKey(id)) {
            return states.get(id);
        } else {
            throw new InvalidGameIdException();
        }
    }

    private Game getGame(int id) {
        if (games.containsKey(id)) {
            return games.get(id);
        } else {
            throw new InvalidGameIdException();
        }
    }

    public Galaxy getGalaxy(int id) {
        return getGame(id).getGalaxy();
    }

    // TODO: Make this check parameters as well
    public boolean isActionValid(int id, String action, String body) {
        GameState state = getState(id);
        List<Method> actions = state.getActions();
        for (Method m : actions) {
            if (m.getName().equals(action)) {
                return true;
            }
        }
        return false;
    }

    public GameState action(int id, String action, String body) {
        GameState state = getState(id);
        List<Method> actions = state.getActions();
        for (Method m : actions) {
            if (m.getName().equals(action)) {
                Object[] parameters = getParametersForMethod(body, m);
                try {
                    GameState previousState = state;
                    state = (GameState)m.invoke(state, parameters);
                    while (state != previousState) { // Some states are just transition states.
                        previousState = state;
                        state = state.init();
                    }
                    return state;
                } catch (IllegalAccessException e) {
                    logger.error("IllegalAccessException? Dang.");
                } catch (InvocationTargetException e) {
                    logger.error("InvocationTargetException... I guess");
                }
            }
        }
        logger.error("Could not find action " + action);
        return state;
    }

    private Object[] getParametersForMethod(String body, Method m) {
        if (body.length() == 0) {
            return new Object[0];
        }
        JsonObject json = new JsonParser().parse(body).getAsJsonObject();
        if (!json.has("parameters")) {
            logger.warn("Passed in JSON for action with content but no parameters");
            return new Object[0];
        }

        JsonObject userParameters = json.getAsJsonObject("parameters");
        List<Object> returnedParameters = new ArrayList<>();
        for (Parameter parameter : m.getParameters()) {
            String name = parameter.getName();
            String type = parameter.getType().getName();
            if (userParameters.has(name)) {
                returnedParameters.add(paramToType(name, type));
            }
        }
        return returnedParameters.toArray();
    }

    private Object paramToType(String name, String type) {
        if (type.equals("int")) {
            return Integer.parseInt(name);
        } else if (type.equals("String")) {
            return name;
        } else if (type.equals(ShipType.class.getName())) {
            return ShipType.Get(name);
        } else {
            logger.error("Could not decode parameter " + name + " into object of type " + type);
            return "????";
        }
    }

    public ShipTypes getShipTypes() {
        return new ShipTypes();
    }

    public Captain getCaptain(int id) {
        return getGame(id).getCaptain();
    }

    public Bank getBank(int id) {
        return getGame(id).getCaptain().bank;
    }

    public class ShipTypes {
        public ShipType[] types;
        ShipTypes() {
            types = ShipType.values();
        }
    }

    private class InvalidGameIdException extends IllegalArgumentException {

    }
}

File: src/main/kotlin/com/malcolmcrum/spacetrader/GameManager.kt
package com.malcolmcrum.spacetrader

import com.malcolmcrum.spacetrader.controllers.GalaxyController
import com.malcolmcrum.spacetrader.controllers.GameStateController
import com.malcolmcrum.spacetrader.controllers.OnPlanetController
import com.malcolmcrum.spacetrader.model.*
import com.malcolmcrum.spacetrader.nouns.Galaxy
import com.malcolmcrum.spacetrader.nouns.Ship
import com.malcolmcrum.spacetrader.states.*
import org.slf4j.LoggerFactory
import java.util.*

private val log = LoggerFactory.getLogger(GameManager::class.java)!!

private const val MAX_SKILL = 10
private const val MAX_INITIAL_SKILL_TOTAL = 20

class GameManager {
    val games = HashMap<GameId, Game>()

    private val galaxyGenerator = GalaxyController()

    fun newGame(commanderName: String, pilot: Int, fighter: Int, trader: Int, engineer: Int, difficulty: Difficulty): GameId {
        assert(pilot <= MAX_SKILL)
        assert(fighter <= MAX_SKILL)
        assert(trader <= MAX_SKILL)
        assert(engineer <= MAX_SKILL)
        assert(pilot + fighter + trader + engineer <= MAX_INITIAL_SKILL_TOTAL)

        val systems = galaxyGenerator.generateGalaxy()
        galaxyGenerator.placeMercenaries(systems)
        galaxyGenerator.placeSpecialEvents(systems)
        galaxyGenerator.setPrices(systems, difficulty)
        galaxyGenerator.setAmounts(systems, difficulty)
        val galaxy = Galaxy(systems, 0, difficulty)
        val captain = CrewMember(commanderName, trader, fighter, engineer, pilot, null)
        val ship = Ship(ShipType.GNAT, captain)
        val id = createGameId()
        val startingSystem = galaxyGenerator.findStartingSystem(ship.type, systems)
        val game = Game(galaxy, id, difficulty, ship, GameState.StartGame(startingSystem))
        games.put(id, game)

        log.info("Created game $id for player $commanderName")
        return id
    }

    fun getController(gameId: GameId): GameStateController {
        val game = games[gameId] ?: throw Exception("No game found for id $gameId")
        val state = game.state
        return when(state) {
            is GameState.GameOver -> TODO()
            is GameState.StartGame -> OnPlanetController(state.startingSystem, game.finances, game.difficulty, game.ship, game.escapePod, game.policeRecord, game.insurance)
            is GameState.OnPlanet -> OnPlanetController(state.system, game.finances, game.difficulty, game.ship, game.escapePod, game.policeRecord, game.insurance)
            is GameState.Travel -> {
                val opponentGenerator = OpponentGenerator(game.difficulty, game.currentWorth(), game.policeRecord, state.destination)
                game.state = TravelController(state, game.policeRecord, game.difficulty, game.ship.type, opponentGenerator).warp()
                return getController(gameId)
            }
            is GameState.PoliceInspection -> PoliceInspection(state.opponent, state.travel, game.finances, game.policeRecord, game.ship.hold, game.difficulty, game.currentWorth())
            is GameState.PoliceAttack -> TODO()
            is GameState.LeaveEncounter -> LeaveEncounter(state.opponent, state.travel, state.text)
        }
    }

    private fun createGameId(): GameId {
        return UUID.randomUUID().toString()
    }

    fun get(id: GameId): Game {
        return games[id] ?: throw Exception("No game found for id $id")
    }
}